### PR TITLE
Json sanitization

### PR DESCRIPTION
**Pull request by VassilIordanov**
Add Json sanitization, improve html sanitization. Also, add report submission checks and graphql introspection restrictions for non-admin.

#### Admin changes
- admins now have the graphiql console under the admin menu

- [ ] anet.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [x] graphql schema has changed:
  It is no longer possible to create/update Tags through GraphQL requests (this functionality was never actually used).

### Checklist
  - [x] Described the user behavior in PR body
  - [ ] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [x] Added and/or updated unit tests
  - [ ] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [x] Opened debt issues for anything not resolved here

### CODE REVIEW

File: src/main/java/mil/dds/anet/resources/ReportResource.java [hunk 1]
```diff
@@ -374,7 +376,19 @@ public Report submitReport(@GraphQLRootContext Map<String, Object> context,
     logger.debug("Attempting to submit report {}, which has advisor org {} and primary advisor {}",
         r, r.getAdvisorOrg(), r.getPrimaryAdvisor());
 
-    // TODO: this needs to be done by either the Author, a Superuser for the AO, or an Administrator
+    if (!r.getAuthor().getUuid().equals(user.getUuid())
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/resources/ReportResource.java:379]: Safer to use:
```
    if (!Objects.equals(r.getAuthorUuid(), user.getUuid())
```

File: src/main/java/mil/dds/anet/resources/ReportResource.java [hunk 2]
```diff
@@ -374,7 +376,19 @@ public Report submitReport(@GraphQLRootContext Map<String, Object> context,
...
+        && !AuthUtils.isSuperUserForOrg(user, r.getAdvisorOrgUuid(), true)
+        && !AuthUtils.isAdmin(user)) {
+      throw new WebApplicationException(
+          "Cannot submit report unless the report's author, his/her super user or an admin",
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/resources/ReportResource.java:383]: Suggested change: `"Cannot submit report unless you are the report's author, his/her super user or an admin"`

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 3]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
     if (input == null) {
       return null;
     }
-    return POLICY_DEFINITION.sanitize(input);
+    return HTML_POLICY_DEFINITION.sanitize(input);
   }
 
+  public static String sanitizeJSON(String inputJson) throws JsonProcessingException {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:222]: Rename method to `sanitizeJson`.
Add a `testSanitizeJson` to `src/test/java/mil/dds/anet/utils/UtilsTest.java`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 4]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+    String sanitizedJson = JsonSanitizer.sanitize(inputJson);
+    JsonNode jsonTree = mapper.readTree(sanitizedJson);
+    internalSanitizeJSONforHTML(jsonTree);
+    return mapper.writeValueAsString(jsonTree);
+  }
+
+  private static void internalSanitizeJSONforHTML(JsonNode jsonNode) {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:229]: Rename method to `internalSanitizeJsonForHtml`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 5]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
     if (input == null) {
       return null;
     }
-    return POLICY_DEFINITION.sanitize(input);
+    return HTML_POLICY_DEFINITION.sanitize(input);
   }
 
+  public static String sanitizeJSON(String inputJson) throws JsonProcessingException {
+    String sanitizedJson = JsonSanitizer.sanitize(inputJson);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:223]: Make `final`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 6]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+    JsonNode jsonTree = mapper.readTree(sanitizedJson);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:224]: Make `final`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 7]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+    internalSanitizeJSONforHTML(jsonTree);
+    return mapper.writeValueAsString(jsonTree);
+  }
+
+  private static void internalSanitizeJSONforHTML(JsonNode jsonNode) {
+    if (jsonNode.isObject()) {
+      ObjectNode objectNode = (ObjectNode) jsonNode;
+      Iterator<Map.Entry<String, JsonNode>> iter = objectNode.fields();
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:232]: Make `final`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 8]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+      while (iter.hasNext()) {
+        Map.Entry<String, JsonNode> entry = iter.next();
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:234]: Make `final`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 9]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+        if (entry.getValue().isTextual()) {
+          objectNode.put(entry.getKey(), sanitizeHtml(entry.getValue().asText()));
+        } else {
+          internalSanitizeJSONforHTML(entry.getValue());
+        }
+        String sanitizedKey = sanitizeHtml(entry.getKey());
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:240]: Make `final`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 10]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+        if (!entry.getKey().equals(sanitizedKey)) {
+          objectNode.remove(entry.getKey());
+          objectNode.set(sanitizedKey, entry.getValue());
+        }
+      }
+    } else if (jsonNode.isArray()) {
+      ArrayNode arrayNode = (ArrayNode) jsonNode;
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:247]: Make `final`.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 11]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+        Map.Entry<String, JsonNode> entry = iter.next();
+        if (entry.getValue().isTextual()) {
+          objectNode.put(entry.getKey(), sanitizeHtml(entry.getValue().asText()));
+        } else {
+          internalSanitizeJSONforHTML(entry.getValue());
+        }
+        String sanitizedKey = sanitizeHtml(entry.getKey());
+        if (!entry.getKey().equals(sanitizedKey)) {
+          objectNode.remove(entry.getKey());
+          objectNode.set(sanitizedKey, entry.getValue());
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:243]: Does `entry.getValue()` here get the value possibly modified above?

File: src/main/java/mil/dds/anet/views/AbstractCustomizableAnetBean.java [hunk 12]
```diff
@@ -17,4 +24,15 @@ public String getCustomFields() {
   public void setCustomFields(String customFields) {
     this.customFields = Utils.trimStringReturnNull(customFields);
   }
+
+  public void checkAndFixCustomFields() {
+    try {
+      setCustomFields(Utils.sanitizeJSON(getCustomFields()));
+    } catch (JsonProcessingException e) {
+      setCustomFields(null);
+      logger.error("Unable to process Json, CustomFields payload discarded", e);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/views/AbstractCustomizableAnetBean.java:33]: Small suggestion: `"Unable to process Json, customFields payload discarded"` as the field is called `customFields` (with lowercase 'c') in the GraphQL schema.

File: src/main/java/mil/dds/anet/resources/ReportResource.java [hunk 13]
```diff
@@ -374,7 +376,19 @@ public Report submitReport(@GraphQLRootContext Map<String, Object> context,
...
+          Status.FORBIDDEN);
+    }
+
+    if (r.getState() != Report.ReportState.DRAFT && r.getState() != Report.ReportState.REJECTED) {
+      throw new WebApplicationException("Can't submit report unless it is either Draft or Rejected",
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/resources/ReportResource.java:388]: `Can't` ⇒ `Cannot` (as above and in other error messages)

File: src/main/java/mil/dds/anet/utils/Utils.java (hunk 11 above)
Comment (VassilIordanov) [reply src/main/java/mil/dds/anet/utils/Utils.java:243]: good catch

Comment (gjvoosten) [reply src/main/java/mil/dds/anet/utils/Utils.java:243]: Be sure to add a test case for it!

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 14]
```diff
@@ -207,9 +216,44 @@ public static String sanitizeHtml(String input) {
...
+    final String sanitizedJson = JsonSanitizer.sanitize(inputJson);
+    final JsonNode jsonTree = mapper.readTree(sanitizedJson);
+    internalSanitizeJsonforHtml(jsonTree);
+    return mapper.writeValueAsString(jsonTree);
+  }
+
+  private static void internalSanitizeJsonforHtml(JsonNode jsonNode) {
+    if (jsonNode.isObject()) {
+      final ObjectNode objectNode = (ObjectNode) jsonNode;
+      for (Iterator<Map.Entry<String, JsonNode>> entryIter = objectNode.fields(); entryIter
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:232]: Iterator can be `final`.

File: src/main/java/mil/dds/anet/resources/GraphQlResource.java [hunk 15]
```diff
@@ -7,7 +7,9 @@
 import graphql.ExecutionResult;
 import graphql.GraphQL;
 import graphql.GraphQLError;
+import graphql.schema.GraphQLCodeRegistry;
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/resources/GraphQlResource.java:10]: Import is unused.

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 16]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
...
+    } else if (jsonNode.isArray()) {
+      final ArrayNode arrayNode = (ArrayNode) jsonNode;
+      for (int i = 0; i < arrayNode.size(); i++) {
+        if (arrayNode.get(i).isTextual()) {
+          final String sanitizedValue = sanitizeHtml(arrayNode.get(i).asText());
+          arrayNode.remove(i);
+          arrayNode.insert(i, sanitizedValue);
+        } else {
+          internalSanitizeJsonForHtml(arrayNode.get(i));
+        }
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:254]: Proposed simplification:
```
@@ -246,9 +246,7 @@ public class Utils {
       final ArrayNode arrayNode = (ArrayNode) jsonNode;
       for (int i = 0; i < arrayNode.size(); i++) {
         if (arrayNode.get(i).isTextual()) {
-          final String sanitizedValue = sanitizeHtml(arrayNode.get(i).asText());
-          arrayNode.remove(i);
-          arrayNode.insert(i, sanitizedValue);
+          arrayNode.set(i, arrayNode.textNode(sanitizeHtml(arrayNode.get(i).asText())));
         } else {
           internalSanitizeJsonForHtml(arrayNode.get(i));
         }
```

File: src/main/java/mil/dds/anet/utils/Utils.java [hunk 17]
```diff
@@ -207,9 +216,45 @@ public static String sanitizeHtml(String input) {
     if (input == null) {
       return null;
     }
-    return POLICY_DEFINITION.sanitize(input);
+    return HTML_POLICY_DEFINITION.sanitize(input);
   }
 
+  public static String sanitizeJson(String inputJson) throws JsonProcessingException {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/utils/Utils.java:222]: Add tests for this method to `UtilsTest`.

File: src/test/java/mil/dds/anet/test/resources/GraphQlResourceTest.java [hunk 18]
```diff
@@ -30,7 +30,48 @@
       LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
   @Test
-  public void test() {
+  public void testIntrospection() {
+    final Map<String, Object> introspectionQuery = new HashMap<String, Object>();
+    introspectionQuery.put("operationName", "IntrospectionQuery");
+    introspectionQuery.put("variables", ImmutableMap.of());
+    introspectionQuery.put("query", "query IntrospectionQuery {"
+        + " __schema { queryType { name } mutationType { name } subscriptionType { name }"
```
Comment (VassilIordanov) [inline src/test/java/mil/dds/anet/test/resources/GraphQlResourceTest.java:38]: consider `import { introspectionQuery } from 'graphql'`

Comment (gjvoosten) [reply src/test/java/mil/dds/anet/test/resources/GraphQlResourceTest.java:38]: In Java code?

Comment (VassilIordanov) [reply src/test/java/mil/dds/anet/test/resources/GraphQlResourceTest.java:38]: indeed :)

Comment (gjvoosten) [reply src/test/java/mil/dds/anet/test/resources/GraphQlResourceTest.java:38]: The suggestion, however, was valid, as there appears to be a `graphql.introspection.IntrospectionQuery` in graphql-java:
`    introspectionQuery.put("query", IntrospectionQuery.INTROSPECTION_QUERY);`

**Pull request merged**